File: python/array_search/uniform_binary_search_production.py
from dataclasses import dataclass


@dataclass(frozen=True)
class UniformBinarySearchResult:
    found: bool
    index: int
    comparisons: int


def build_step_table(length):
    if length <= 0:
        return []

    largest = 1
    while largest <= length // 2:
        largest *= 2

    steps = []
    step = largest
    while step >= 1:
        steps.append(step)
        if step == 1:
            break
        step //= 2
    return steps


def uniform_binary_search(values, target, *, start=0, stop=None):
    length = len(values)
    end = length if stop is None else stop

    if start < 0 or start > length or end < start or end > length:
        raise ValueError("range must satisfy 0 <= start <= stop <= len(values)")

    range_length = end - start
    base = -1
    comparisons = 0

    for step in build_step_table(range_length):
        probe = base + step
        if probe >= range_length:
            continue

        index = start + probe
        comparisons += 1

        if values[index] == target:
            return UniformBinarySearchResult(True, index, comparisons)
        if values[index] < target:
            base = probe

    return UniformBinarySearchResult(False, -1, comparisons)

File: java/array_search/UniformBinarySearchProduction.java
public final class UniformBinarySearchProduction {
    public record SearchResult(boolean found, int index, int comparisons) {
    }

    private UniformBinarySearchProduction() {
    }

    public static SearchResult search(int[] values, int target) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }
        return search(values, target, 0, values.length);
    }

    public static SearchResult search(int[] values, int target, int start, int stop) {
        if (values == null) {
            throw new IllegalArgumentException("values must not be null");
        }
        validateRange(values.length, start, stop);

        int rangeLength = stop - start;
        int largest = 1;
        while (largest <= rangeLength / 2) largest *= 2;

        int base = -1;
        int comparisons = 0;

        for (int step = rangeLength == 0 ? 0 : largest; step >= 1; step /= 2) {
            int probe = base + step;
            if (probe >= rangeLength) continue;

            int index = start + probe;
            comparisons += 1;

            if (values[index] == target) {
                return new SearchResult(true, index, comparisons);
            }
            if (values[index] < target) base = probe;
        }

        return new SearchResult(false, -1, comparisons);
    }

    private static void validateRange(int length, int start, int stop) {
        if (start < 0 || start > length || stop < start || stop > length) {
            throw new IllegalArgumentException("range must satisfy 0 <= start <= stop <= values.length");
        }
    }
}
